dmabuf v4 bump broke build

### mpv version

Issue persists on 89bd6ead6c6f710c42ce2f6f4c5abdb157090761 (current master) and bisecting points to 666cb91cf12a4f8b42087530104513b0fc4ba16d as the first bad commit.

### Reproduction steps

`./waf configure --disable-pipewire && ./waf build`

As attested by the CI still working, this doesn't seem to affect all system configurations. Here, all wayland and dmabuf related bits should come from Debian Bullseye.
See config log for details: [waf_config.log.txt](https://github.com/mpv-player/mpv/files/9894149/waf_config.log.txt)
*(PipeWire is disabled to work around #10786, which was only fixed after 666cb91cf12a4f8b42087530104513b0fc4ba16d)*

### Workaround

`--disable-wayland`

### Error

```
[181/270] Compiling build/generated/wayland/idle-inhibit-unstable-v1.c
../video/out/wayland_common.c:1128:21: error: variable ‘dmabuf_feedback_listener’ has initializer but incomplete type
 1128 | static const struct zwp_linux_dmabuf_feedback_v1_listener dmabuf_feedback_listener = {
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../video/out/wayland_common.c:1129:5: warning: excess elements in struct initializer
 1129 |     done,
      |     ^~~~
../video/out/wayland_common.c:1129:5: note: (near initialization for ‘dmabuf_feedback_listener’)
../video/out/wayland_common.c:1130:5: warning: excess elements in struct initializer
 1130 |     format_table,
      |     ^~~~~~~~~~~~
../video/out/wayland_common.c:1130:5: note: (near initialization for ‘dmabuf_feedback_listener’)
../video/out/wayland_common.c:1131:5: warning: excess elements in struct initializer
 1131 |     main_device,
      |     ^~~~~~~~~~~
../video/out/wayland_common.c:1131:5: note: (near initialization for ‘dmabuf_feedback_listener’)
../video/out/wayland_common.c:1132:5: warning: excess elements in struct initializer
 1132 |     tranche_done,
      |     ^~~~~~~~~~~~
../video/out/wayland_common.c:1132:5: note: (near initialization for ‘dmabuf_feedback_listener’)
../video/out/wayland_common.c:1133:5: warning: excess elements in struct initializer
 1133 |     tranche_target_device,
      |     ^~~~~~~~~~~~~~~~~~~~~
../video/out/wayland_common.c:1133:5: note: (near initialization for ‘dmabuf_feedback_listener’)
../video/out/wayland_common.c:1134:5: warning: excess elements in struct initializer
 1134 |     tranche_formats,
      |     ^~~~~~~~~~~~~~~
../video/out/wayland_common.c:1134:5: note: (near initialization for ‘dmabuf_feedback_listener’)
../video/out/wayland_common.c:1135:5: warning: excess elements in struct initializer
 1135 |     tranche_flags,
      |     ^~~~~~~~~~~~~
../video/out/wayland_common.c:1135:5: note: (near initialization for ‘dmabuf_feedback_listener’)
../video/out/wayland_common.c: In function ‘registry_handle_add’:
../video/out/wayland_common.c:1162:31: error: implicit declaration of function ‘zwp_linux_dmabuf_v1_get_default_feedback’; did you mean ‘zwp_linux_dmabuf_v1_get_user_data’? [-Werror=implicit-function-declaration]
 1162 |         wl->dmabuf_feedback = zwp_linux_dmabuf_v1_get_default_feedback(wl->dmabuf);
      |                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                               zwp_linux_dmabuf_v1_get_user_data
../video/out/wayland_common.c:1162:29: warning: assignment to ‘struct zwp_linux_dmabuf_feedback_v1 *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
 1162 |         wl->dmabuf_feedback = zwp_linux_dmabuf_v1_get_default_feedback(wl->dmabuf);
      |                             ^
../video/out/wayland_common.c:1163:9: error: implicit declaration of function ‘zwp_linux_dmabuf_feedback_v1_add_listener’; did you mean ‘zwp_linux_dmabuf_v1_add_listener’? [-Werror=implicit-function-declaration]
 1163 |         zwp_linux_dmabuf_feedback_v1_add_listener(wl->dmabuf_feedback, &dmabuf_feedback_listener, wl);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |         zwp_linux_dmabuf_v1_add_listener
../video/out/wayland_common.c: At top level:
../video/out/wayland_common.c:1128:59: error: storage size of ‘dmabuf_feedback_listener’ isn’t known
 1128 | static const struct zwp_linux_dmabuf_feedback_v1_listener dmabuf_feedback_listener = {
      |                                                           ^~~~~~~~~~~~~~~~~~~~~~~~
cc1: some warnings being treated as errors

../common/encode_lavc.c: In function ‘encode_lavc_init’:
../common/encode_lavc.c:158:23: warning: assignment discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
  158 |     p->muxer->oformat = ctx->oformat;
      |                       ^
../common/encode_lavc.c: In function ‘find_codec_for’:
../common/encode_lavc.c:752:56: warning: passing argument 1 of ‘av_guess_codec’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
  752 |         codec = avcodec_find_encoder(av_guess_codec(ctx->oformat, NULL,
      |                                                     ~~~^~~~~~~~~
In file included from ../common/encode_lavc.h:28,
                 from ../common/encode_lavc.c:27:
/usr/include/x86_64-linux-gnu/libavformat/avformat.h:2705:58: note: expected ‘AVOutputFormat *’ but argument is of type ‘const AVOutputFormat *’
 2705 | enum AVCodecID av_guess_codec(ff_const59 AVOutputFormat *fmt, const char *short_name,
      |                                          ~~~~~~~~~~~~~~~~^~~
```

--- a/generated/wayland/meson.build
+++ b/generated/wayland/meson.build
@@ -8,6 +8,8 @@ protocols = [[wl_protocol_dir, 'stable/presentation-time/presentation-time.xml']
 wl_protocols_source = []
 wl_protocols_headers = []
 
+features += {'wayland_protocols_1_24': wayland['deps'][2].version().version_compare('>=1.24')}
+
 foreach p: protocols
     xml = join_paths(p)
     wl_protocols_source += custom_target(xml.underscorify() + '_c',

--- a/video/out/wayland_common.c
+++ b/video/out/wayland_common.c
@@ -1094,6 +1094,7 @@ static const struct zwp_linux_dmabuf_v1_listener dmabuf_listener = {
     dmabuf_format
 };
 
+#if HAVE_WAYLAND_PROTOCOLS_1_24
 static void done(void *data,
                  struct zwp_linux_dmabuf_feedback_v1 *zwp_linux_dmabuf_feedback_v1)
 {
@@ -1153,6 +1154,7 @@ static const struct zwp_linux_dmabuf_feedback_v1_listener dmabuf_feedback_listen
     tranche_formats,
     tranche_flags,
 };
+#endif
 
 static void registry_handle_add(void *data, struct wl_registry *reg, uint32_t id,
                                 const char *interface, uint32_t ver)
@@ -1178,10 +1180,11 @@ static void registry_handle_add(void *data, struct wl_registry *reg, uint32_t id
 
     if (!strcmp (interface, zwp_linux_dmabuf_v1_interface.name) && (ver >= 4) && found++) {
         wl->dmabuf = wl_registry_bind(reg, id, &zwp_linux_dmabuf_v1_interface, 4);
+#if HAVE_WAYLAND_PROTOCOLS_1_24
         wl->dmabuf_feedback = zwp_linux_dmabuf_v1_get_default_feedback(wl->dmabuf);
         zwp_linux_dmabuf_feedback_v1_add_listener(wl->dmabuf_feedback, &dmabuf_feedback_listener, wl);
-    }
-    else if (!strcmp (interface, zwp_linux_dmabuf_v1_interface.name) && (ver >= 2) && found++) {
+#endif
+    } else if (!strcmp (interface, zwp_linux_dmabuf_v1_interface.name) && (ver >= 2) && found++) {
         wl->dmabuf = wl_registry_bind(reg, id, &zwp_linux_dmabuf_v1_interface, 2);
         zwp_linux_dmabuf_v1_add_listener(wl->dmabuf, &dmabuf_listener, wl);
         wl->drm_format_ct_max = 64;

--- a/video/out/wayland_common.h
+++ b/video/out/wayland_common.h
@@ -83,7 +83,8 @@ struct vo_wayland_state {
 
     /* linux-dmabuf */
     struct zwp_linux_dmabuf_v1 *dmabuf;
-    struct zwp_linux_dmabuf_feedback_v1 *dmabuf_feedback;
+    /* TODO: unvoid this if required wayland protocols is bumped to 1.24+ */
+    void *dmabuf_feedback;
     void *format_map;
     uint32_t format_size;
     /* TODO: remove these once zwp_linux_dmabuf_v1 version 2 support is removed. */

--- a/wscript
+++ b/wscript
@@ -531,6 +531,11 @@ video_output_features = [
                                  'wayland-cursor', '>= 1.15.0',
                                  'xkbcommon',      '>= 0.3.0'),
     } , {
+        'name': 'wayland-protocols-1-24',
+        'desc': 'wayland-protocols version 1.24+',
+        'deps': 'wayland',
+        'func': check_pkg_config('wayland-protocols >= 1.24'),
+    } , {
         'name': 'memfd_create',
         'desc': "Linux's memfd_create()",
         'deps': 'wayland',

